text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1740 @@
+- Feature Name: `generalized_type_ascription`
+- Start Date: 2018-08-06
+- RFC PR: _
+- Rust Issue: _
+
+# Summary
+[summary]: #summary
+
+[RFC 803]: https://github.com/rust-lang/rfcs/blob/master/text/0803-type-ascription.md
+
+This RFC supersedes and subsumes [RFC 803].
+We finalize a general notion of type ascription uniformly in patterns,
+expressions, `let` bindings, and `fn` definitions. You may now for example write:
+
+```rust
+let x = (0..10).collect() : Vec<_>;
+
+do_stuff(try : Option<u8> { .. });
+
+do_stuff(async : u8 { .. });
+
+do_stuff(unsafe : u8 { .. });
+
+do_stuff(loop : u8 { .. });
+
+let alpha: u8 = expr;
+    ^^^^^^^^^
+
+let [x: u8, y, z] = stuff();
+    ^^^^^^^^^^^^^
+
+if let Some(beta: u8) = expr { .. }
+            ^^^^^^^^
+
+fn foo(Wrapping(alpha: usize)) {}
+       ^^^^^^^^^^^^^^^^^^^^^^
+```
+
+Here, the underlined bits are patterns.
+Note however that this RFC does *not* introduce global type inference.
+
+Finally, we lint (warn-by-default) when a user writes
+`Foo { $field: $ident : $type }`
+and the compiler instead suggests:
+`Foo { $field: ($ident : $type) }`.
+
+# Motivation
+[motivation]: #motivation
+
+## Type ascription is useful
+
+[RFC_803_motivation]: https://github.com/rust-lang/rfcs/blob/master/text/0803-type-ascription.md#motivation
+
+[pointfree persuasion]: https://en.wikipedia.org/wiki/Tacit_programming
+
+[TwoHardThings]: https://martinfowler.com/bliki/TwoHardThings.html
+
+Type ascription is useful. A motivation for the feature is noted in the merged,
+but thus far not stabilized, [RFC 803][RFC_803_motivation] which introduces
+type ascription in expression contexts as `expr : T`. We reinforce that RFC
+with more motivation:
```

Comment:
These are all good motivations, but I think they vary in strength.  As a nit, consider reordering the strongest ones first.  Obviously that's subjective, but right now (1) feels like it'll get "one variable just isn't that big of a deal" push-back.  I really like (4), for example, and (2) is IMHO a stronger phrasing of the annoyance.